### PR TITLE
detect network location change as part of motion sensing

### PR DESCRIPTION
**Pull request by garvankeeley**
Addresses https://github.com/mozilla/MozStumbler/issues/1496

### CODE REVIEW

File: libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java [hunk 1]
```diff
@@ -128,8 +130,54 @@ public void stop() {
...
+        public void start(Context c) {
+            mContext = c;
+            LocationManager lm = (LocationManager) mContext.getSystemService(Context.LOCATION_SERVICE);
+
+            if (mLastLocation == null) {
+                mLastLocation = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER);
+            }
+
+            if (mLastLocation != null) {
+                lm.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, mNetworkLocationListener);
```
Comment (crankycoder) [inline libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java]: Why is this null checking before requesting locations?  I would think it's ok to request new locations even if we didn't have a last known location in the GPS provider.

File: libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java [hunk 2]
```diff
@@ -128,8 +130,54 @@ public void stop() {
...
+
+    static class NetworkLocationChangeDetector {
+        Location mLastLocation;
+        Context mContext;
+        private float DIST_THRESHOLD_M = 30.0f;
+
+        private LocationListener mNetworkLocationListener = new LocationListener() {
+            public void onLocationChanged(Location location) {
+                if (mLastLocation == null) {
+                    return;
```
Comment (crankycoder) [inline libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java]: Shouldn't this do `mLastLocation = location` before returning?

File: libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java [hunk 3]
```diff
@@ -128,8 +130,54 @@ public void stop() {
...
+            public void onProviderEnabled(String provider) {}
+            public void onProviderDisabled(String provider) {}
+        };
+
+        public void start(Context c) {
+            mContext = c;
+            LocationManager lm = (LocationManager) mContext.getSystemService(Context.LOCATION_SERVICE);
+
+            if (mLastLocation == null) {
+                mLastLocation = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER);
```
Comment (crankycoder) [inline libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/MotionSensor.java:166]: Maybe change this to get whatever the highest accuracy location is by doing something like:

``` java
matchingProviders = locationManager.getAllProviders();
for (String provider: matchingProviders) {
   location = locationManager.getLastKnownLocation(provider);

   // check that location is not null and grab the location with the best accuracy 
   // here, then bind it to mLastLocation
}
```

**Comment by crankycoder** [comment]
I've completed the review.  Mostly - I think we should try to just grab any location at first.

**Comment by garvankeeley** [comment]
Note: I did not change the false positive filter to allow for late initialization of its reference location (mLastLocation), that is by-design required to have an initial GPS location and then wait to see if later locations are the same location. Without the initial GPS location, the class won't attempt to filter a false positive, which is correct behaviour.

**Pull request merged**